python/ray/data/impl/null_aggregate.py

```diff
@@ -135,52 +183,43 @@ def _accum(a: AggType, r: T) -> AggType:
     return _accum
 
 
-def _null_wrap_merge(
+def _null_wrap_accumulate_block(
     ignore_nulls: bool,
-    merge: Callable[[AggType, AggType], AggType],
-) -> AggType:
+    accum_block: Callable[[AggType, BlockAccessor[T]], AggType],
+) -> Callable[[AggType, BlockAccessor[T]], AggType]:
     """
-    Wrap merge function with null handling.
-
-    The returned merge function expects a1 and a2 to be either None or of the form:
-    a = [acc_data_1, ..., acc_data_2, has_data].
+    Wrap vectorized aggregate function with null handling.
 
-    This merges two accumulations subject to the following null rules:
-    1. If a1 is empty and a2 is empty, return empty accumulation.
-    2. If a1 (a2) is empty and a2 (a1) is None, return None.
-    3. If a1 (a2) is empty and a2 (a1) is non-None, return a2 (a1).
-    4. If a1 (a2) is None, return a2 (a1) if ignoring nulls, None otherwise.
-    5. If a1 and a2 are both non-null, return merge(a1, a2).
+    This performs a block accumulation subject to the following null rules:
+    1. If any row is null and ignore_nulls=False, return None.
+    2. If at least one row is not null and ignore_nulls=True, return the block
+       accumulation.
+    3. If all rows are null and ignore_nulls=True, return the base accumulation.
+    4. If all rows non-null, return the block accumulation.
 
     Args:
         ignore_nulls: Whether nulls should be ignored or cause a None result.
-        merge: The core merge function to wrap.
+        accum_block: The core vectorized aggregate function to wrap.
 
     Returns:
-        A new merge function that handles nulls.
+        A new vectorized aggregate function that handles nulls.
     """
 
-    def _merge(a1: AggType, a2: AggType) -> AggType:
-        if a1 is None:
-            # If we're ignoring nulls, propagate a2; otherwise, propagate None.
-            return a2 if ignore_nulls else None
-        unwrapped_a1, a1_has_data = _unwrap_acc(a1)
-        if not a1_has_data:
-            # If a1 is empty, propagate a2.
-            # No matter whether a2 is a real value, empty, or None,
-            # propagating each of these is correct if a1 is empty.
-            return a2
-        if a2 is None:
-            # If we're ignoring nulls, propagate a1; otherwise, propagate None.
-            return a1 if ignore_nulls else None
-        unwrapped_a2, a2_has_data = _unwrap_acc(a2)
-        if not a2_has_data:
-            # If a2 is empty, propagate a1.
-            return a1
-        a = merge(unwrapped_a1, unwrapped_a2)
-        return _wrap_acc(a, has_data=True)
+    def _accum_block_null(a: AggType, block_acc: BlockAccessor[T]) -> AggType:
+        a_data, _ = _unwrap_acc(a)
```

Comment:
>what if a is already None?

`a` is expected to come from one of the `_null_wrap_*` helpers, so it will adhere to this structure.

>I feel this might be a simpler impl:
>
>accm_block takes a block and returns a aggregator, we then merge the returned aggregator with the input aggregator using the existing _null_wrap_merge so we share the same merge logic with null handling?

Are you referring to going away from the top-level accumulator API? IIUC, you're suggesting that we change `accumulate_block`'s type from `Callable[[AggType, Block[T]], AggType]` to `Callable[[Block[T]], AggType]`, and apply `merge`. But I'm not sure how we would make the `accumulate_row: Callable[[AggType, T], AggType]` API fit in here.

Reply:
Or are you instead suggesting the `_null_wrap_accumulate_block` implementation should change to leverage `_null_wrap_merge`? That seems doable.